### PR TITLE
Partykit opt out

### LINKED ISSUE #10014: Party kit notif and opt out
image

## Design
- in drop page handoff: https://www.figma.com/file/JXNtryaNYhY0XJNArzvPQx/Handoff?type=design&node-id=394-6126&mode=design&t=YQJR93nSuIPTEKpt-11

easy modal design
image

- opt out button
- dive in button - closes the modal and give consent 
- the default is always on, only the opt out changes anything 
- closing the modals keeps the partykit on 

## Learn more
- leads to https://hello.kodadot.xyz/tutorial/generative-art/party-kit
- can you please update the hello so that it mentions it can be turned off in the settings @JustLuuuu 

## Settings
**since the modal will only display once - first time the user enters our drop page - he can change the preference in the settings** 
image
- there is partykit heading added, tooltip with on/off button
image

please change the heading styling of "user interface mode" to black bold text like it is in the designs

### PR DESCRIPTION
**Pull request by iMac7**
**Thank you for your contribution** to the [Koda - Generative Art Marketplace](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring

## Context

- [x] Closes #10014 
- [ ] Requires deployment <worker/stick/speck>

#### Before submitting pull request, please make sure:

- [ ] My contribution builds **clean without any errors or warnings**
- [ ] I've merged recent default branch -- **main** and I've no conflicts
- [ ] I've tried to respect high code quality standards
- [ ] I've didn't break any original functionality

#### Optional

- [ ] I've tested it at </ahp/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Did your issue had any of the "$" label on it?

- [x] Fill up your DOT address: [Payout](https://canary.kodadot.xyz/dot/transfer?target=14dLrAvjDw7cFLaxM7VWR6YtHd7bGZY2QpzXDhAJJ4yv6oGw&usdamount=undefined)

#### Community participation

- [x] [Are you at Koda Ecosystem Telegram?](https://t.me/koda_eco)

## Screenshot 📸

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.
Screenshot_20240408_012242

### CODE REVIEW

File: components/collection/drop/GenerativeLayout.vue [hunk 1]
```diff
@@ -74,6 +76,8 @@ import {
 } from '@/components/drops/useDrops'
 import { useCollectionActivity } from '@/composables/collectionActivity/useCollectionActivity'
 import { useCollectionMinimal } from '@/components/collection/utils/useCollectionDetails'
+import PartyModal from '@/components/common/party/PartyModal.vue'
```
Comment (hassnian) [inline components/collection/drop/GenerativeLayout.vue:79]:
```suggestion

```

can be auto imported. 

CleanShot 2024-04-08 at 10 00 05@2x

File: components/collection/drop/GenerativeLayout.vue [hunk 2]
```diff
@@ -84,6 +88,7 @@ const { drop } = useDrop()
 const { previewItem } = storeToRefs(useDropStore())
 const { mintedAmountForCurrentUser, description } = useCollectionEntity()
 const { width } = useWindowSize()
+const preferencesStore = usePreferencesStore()
```
Comment (hassnian) [inline components/collection/drop/GenerativeLayout.vue:91]:
```suggestion
const { firstTimePartyMode } = storeToRefs(usePreferencesStore())
```
inside `PartyModal`

File: components/common/party/PartyModal.vue [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<template>
+  <div
+    class="party shadow-primary h-fit fixed !top-20 md:!top-24 sm:left-4 border-2 bg-background-color p-2 sm:!p-6 flex justify-between gap-2 sm:gap-6 max-w-[498px]">
+    <div class="flex-1 min-w-[91px] h-[116px] flex justify-center items-center">
+      <img :src="'/drop/party.svg'" />
+    </div>
```
Comment (hassnian) [inline components/common/party/PartyModal.vue:6]: atm we don't get any kind of animation. it's pretty abrupt, check `CookieBanner.vue`

CleanShot 2024-04-08 at 10 09 52@2x

File: components/common/party/PartyModal.vue [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+const { $i18n } = useNuxtApp()
+const preferencesStore = usePreferencesStore()
+
+const closeModal = () => {
+  preferencesStore.firstTimePartyMode = false
+}
+const stopParty = () => {
+  preferencesStore.partyMode = false
+  closeModal()
+}
```
Comment (hassnian) [inline components/common/party/PartyModal.vue:46]: it should only be shown **once**

CleanShot 2024-04-08 at 10 13 31@2x


CleanShot 2024-04-08 at 10 01 10

File: composables/party/useCursorParty.ts [hunk 5]
```diff
@@ -97,5 +100,5 @@ export default ({
     })
   })
 
-  return { connections: visibleConnections }
+  return { connections: visibleConnections, closeConnection }
```
Comment (hassnian) [inline composables/party/useCursorParty.ts:103]:
```suggestion
  return { connections: visibleConnections }
```

File: composables/party/useParty.ts [hunk 6]
```diff
@@ -57,23 +59,25 @@ export default <T>({ room, onMessage }: UsePartyParams<T>) => {
     wss.value.get(room.value)?.send(JSON.stringify(value))
   }
 
+  const closeConnection = () => {
```
Comment (hassnian) [inline composables/party/useParty.ts:62]:
```suggestion
  const closeAllConnections = () => {
```

File: composables/party/useParty.ts [hunk 7]
```diff
@@ -57,23 +59,25 @@ export default <T>({ room, onMessage }: UsePartyParams<T>) => {
...
+      ws.close()
+      wss.value.delete(room)
+    }
+  }
+
   watch(
     room,
     (value) => {
-      if (value) {
+      if (value && preferencesStore.partyMode) {
```
Comment (hassnian) [inline composables/party/useParty.ts:73]: `useParty.ts` it's a generic composable , that might be used for other party types 

inside `useCursorParty` do something like 

CleanShot 2024-04-08 at 10 24 58@2x

and you can 
```suggestion
      if (value && !disabled) {
```

File: stores/preferences.ts [hunk 8]
```diff
@@ -164,6 +168,12 @@ export const usePreferencesStore = defineStore('preferences', {
     setLayoutClass(payload) {
       this.layoutClass = payload
     },
+    setPartyMode(payload) {
```
Comment (hassnian) [inline stores/preferences.ts:171]: add param type

File: stores/preferences.ts [hunk 9]
```diff
@@ -164,6 +168,12 @@ export const usePreferencesStore = defineStore('preferences', {
...
+      this.partyMode = payload
+    },
+    setFirstTimePartyMode(payload) {
```
Comment (hassnian) [inline stores/preferences.ts:174]: add param type

File: locales/en.json [hunk 10]
```diff
@@ -1552,6 +1555,12 @@
       "public": "Public Mint"
     }
   },
+  "partyModal": {
+    "title": "Experience The Drop Party",
+    "text": "See and join other collectors browsing the drop in real time.",
+    "no": "OPT OUT",
+    "yes": "DIVE IN!"
```
Comment (hassnian) [inline locales/en.json:1562]: match the design pls

CleanShot 2024-04-08 at 10 31 11@2x

File: composables/party/useParty.ts [hunk 11]
```diff
@@ -57,23 +59,25 @@ export default <T>({ room, onMessage }: UsePartyParams<T>) => {
     wss.value.get(room.value)?.send(JSON.stringify(value))
   }
 
+  const closeConnection = () => {
+    for (const [room, ws] of wss.value) {
+      console.log(`[PARTY::CONNECTION] Closing connection with room ${room}`)
+      ws.close()
+      wss.value.delete(room)
```
Comment (hassnian) [inline composables/party/useParty.ts:66]: make method closeConnection

**Review by hassnian — CHANGES_REQUESTED** [review]

File: components/common/party/PartyModal.vue (hunk 3 above)
Comment (iMac7) [reply components/common/party/PartyModal.vue]: I have added a <transition> wrapper but I can't see an animation on the cookiebanner as well :thinking:

File: composables/party/useParty.ts [hunk 12]
```diff
@@ -57,23 +60,37 @@ export default <T>({ room, onMessage }: UsePartyParams<T>) => {
...
-  })
+  watch(
+    () => preferencesStore.partyMode,
+    (newValue) => {
+      if (newValue === false) {
+        closeAllConnections()
+      }
+    },
+  )
+
```
Comment (hassnian) [inline composables/party/useParty.ts:92]:
```suggestion
  watch(
    disabled,
    (isDisabled) => {
      if (isDisabled) {
        closeAllConnections()
      }
    },
  )

```

`useParty` needs to be generic, trigger this logic via the param , make it a computed and react here to it 

remove the store

File: composables/party/useParty.ts [hunk 13]
```diff
@@ -1,15 +1,18 @@
...
+  disabled?: boolean
 }
 
 const PARTY_SOCKET_HOST = 'https://partykit.kodadot.workers.dev/'
 const wss = ref(new Map<string, PartySocket>())
 
-export default <T>({ room, onMessage }: UsePartyParams<T>) => {
+export default <T>({ room, onMessage, disabled }: UsePartyParams<T>) => {
   const { accountId } = useAuth()
+  const preferencesStore = usePreferencesStore()
```
Comment (hassnian) [inline composables/party/useParty.ts:15]:
```suggestion

```

File: components/common/party/PartyModal.vue [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+</template>
+
+<script setup lang="ts">
+import { NeoButton, NeoIcon } from '@kodadot1/brick'
+import { usePreferencesStore } from '@/stores/preferences'
+
+const { $i18n } = useNuxtApp()
+const preferencesStore = usePreferencesStore()
+
+const emit = defineEmits(['close'])
```
Comment (hassnian) [inline components/common/party/PartyModal.vue:43]:
```suggestion
const emit = defineEmits(['close'])

const { $i18n } = useNuxtApp()
const preferencesStore = usePreferencesStore()
```

File: components/collection/drop/GenerativeLayout.vue [hunk 15]
```diff
@@ -84,6 +87,8 @@ const { drop } = useDrop()
 const { previewItem } = storeToRefs(useDropStore())
 const { mintedAmountForCurrentUser, description } = useCollectionEntity()
 const { width } = useWindowSize()
+const { firstTimePartyMode } = usePreferencesStore()
+const isPartyModalOpen = ref(firstTimePartyMode)
```
Comment (hassnian) [inline components/collection/drop/GenerativeLayout.vue:91]:
```suggestion
const isPartyModalOpen = ref(usePreferencesStore().firstTimePartyMode)
```

**Review by hassnian — CHANGES_REQUESTED** [review]

**Comment by exezbcz** [comment]
image
image

find two differences 

hint: its the padding, layout and capital text

**Comment by iMac7** [comment]
@exezbcz I reduced the padding a bit on mobile because the buttons are too close otherwise and the text wraps too much. Any modifications to this?

image

**Comment by exezbcz** [comment]
@iMac7 sure, will comment soon

**Review by hassnian — APPROVED** [review]

**Comment by hassnian** [comment]
@iMac7 pls don't resolve any conversation, let the reviewer decide if it's resolved or not :D

thanks

**Comment by hassnian** [comment]
> > @iMac7 pls don't resolve any conversation, let the reviewer decide if it's resolved or not :D
> 
> Right, there were just too many XD

I can understand, just collapse them , add a comment or react to them to keep track of the ones that you've completed

**Comment by exezbcz** [comment]
helo sir, sorry for the late response, this is the layout for mobile drops 

image


updated in figma: https://www.figma.com/file/JXNtryaNYhY0XJNArzvPQx/Handoff?type=design&node-id=394-6126&mode=design&t=nXpA2m1dD9KC6XOO-11

**Review by vikiival — COMMENTED** [review]
Waiting for mobile version

**Comment by exezbcz** [comment]
this does not look right 
image



image



padding to 24px 
image



and the graphics bigger please


image



mobile looks okay apart from the padding

**Comment by exezbcz** [comment]
helo sir, how do i test it now? it does not show for me now? 


One thing i would like to make sure as well: 
- it will display to the user everytime he enters the drop until:
	- he closes it - its "opt in"
	- clicks dive in - opt in
	- clicks opt out - partykit off - opt out

File: components/collection/drop/GenerativeLayout.vue [hunk 16]
```diff
@@ -64,6 +64,8 @@
   <CollectionDropCursorParty
     :drop-alias="drop.alias"
     :user-minted-count="mintedAmountForCurrentUser" />
+
+  <PartyModal v-if="isPartyModalOpen" @close="isPartyModalOpen = false" />
```
Comment (vikiival) [inline components/collection/drop/GenerativeLayout.vue:68]: shouldn't `@close` update you prefference store?

**Review by vikiival — COMMENTED** [review]
needs clarification

**Comment by prury** [comment]
Performed some tests and even some dev comments PR does not deliver basic functionality

what i have detected: 

- opted in in two different browsers sessions and party kit does not work: 
[pkitvideo.webm](https://github.com/kodadot/nft-gallery/assets/36627808/66b10e6c-e614-4f0f-ae75-499609613b96)

- responsivity issues on mobile: [pkit1](https://github.com/kodadot/nft-gallery/assets/36627808/713f25b4-fd5b-4bf4-802a-c9d5ae48cb11)
- z-index issues: [pkit2](https://github.com/kodadot/nft-gallery/assets/36627808/ebe585ad-2646-4f1a-9c95-8a645d9ab716)
- cursor not changing when hovering on close icon: [pkit-no-cursor](https://github.com/kodadot/nft-gallery/assets/36627808/8d371e6f-3975-4387-a9b0-116a233922e6)

Closing PR

**Pull request closed without merging**